EXILED/Exiled.Events/Patches/Events/Player/Shooting.cs

```diff
@@ -39,34 +34,59 @@ private static IEnumerable<CodeInstruction> Transpiler(IEnumerable<CodeInstructi
             List<CodeInstruction> newInstructions = ListPool<CodeInstruction>.Pool.Get(instructions);
 
             Label returnLabel = generator.DefineLabel();
-
-            int offset = 1;
-            int index = newInstructions.FindIndex(instruction => instruction.opcode == OpCodes.Stloc_2) + offset;
+            Label continueLabel = generator.DefineLabel();
+
+            /*
+            [] <= Here
+            IL_0078: ldarg.2      // processingMethod
+            IL_0079: ldarg.0      // this
+            IL_007a: ldfld        class ReferenceHub InventorySystem.Items.Firearms.Modules.Misc.ShotBacktrackData::PrimaryTargetHub
+            IL_007f: callvirt     instance void class [mscorlib]System.Action`1<class ReferenceHub>::Invoke(!0/*class ReferenceHub* /)
+             */
+            int hasTargetIndex = newInstructions.FindIndex(instruction => instruction.IsLdarg(2));
+            List<Label> hasTargetLabels = newInstructions[hasTargetIndex].ExtractLabels();
+
+            /*
+            [] <= Here
+            IL_0092: ldarg.2      // processingMethod
+            IL_0093: ldnull       // null
+            IL_0094: callvirt     instance void class [mscorlib]System.Action`1<class ReferenceHub>::Invoke(!0/*class ReferenceHub* /)
+             */
+            int noTargetIndex = newInstructions.FindIndex(hasTargetIndex + 1, instruction => instruction.IsLdarg(2));
+            List<Label> noTargetLabels = newInstructions[noTargetIndex].ExtractLabels();
+
+            ConstructorInfo constructorInfo = Constructor(
+                typeof(ShootingEventArgs),
+                new[] { typeof(InventorySystem.Items.Firearms.Firearm), typeof(ShotBacktrackData).MakeByRefType() });
+
+            CodeInstruction[] patchInstructions =
+            {
+                // ShootingEventArgs ev = new(firearm, this)
+                new(OpCodes.Ldarg_1),
+                new(OpCodes.Ldarg_0),
+                new(OpCodes.Newobj, constructorInfo),
+
+                // Handlers.Player.OnShooting(ev)
+                new(OpCodes.Dup), // Dup to keep ev on the stack
+                new(OpCodes.Call, Method(typeof(Handlers.Player), nameof(Handlers.Player.OnShooting))),
+
+                // if (!ev.IsAllowed) return
+                new(OpCodes.Callvirt, PropertyGetter(typeof(ShootingEventArgs), nameof(ShootingEventArgs.IsAllowed))),
+                new(OpCodes.Brtrue_S, continueLabel),
+                new(OpCodes.Leave_S, returnLabel),
+
+                new CodeInstruction(OpCodes.Nop).WithLabels(continueLabel),
```

Comment:
my main worry is that you insert this il code 2 times, but use 1 label `continueLabel` the same way, I highly doubt it works as intended

Reply:
or harmony handles those things? idk, ill test

Reply:
Now because you use one array for 2 insertions, all labels gets doubled😭😭😭
<img width="185" alt="image" src="https://github.com/user-attachments/assets/8835e93c-52be-48c7-8a2b-3cbe1e81435f">
<img width="181" alt="image" src="https://github.com/user-attachments/assets/70fcc56c-ac9d-4113-a535-f8d5fbf8ba55">
<img width="139" alt="image" src="https://github.com/user-attachments/assets/d96639fc-a968-4764-b017-0a38662ae04b">
<img width="131" alt="image" src="https://github.com/user-attachments/assets/d2a1751b-263c-40b8-b44f-9887070be2d8">

Reply:
You can create method like
```cs
private IEnumerable<CodeInstruction> GetInstructions(Label continueLabel)
{
    yield return ...
    ....
}
```
that have labels or/and startInstruction as arguments, for setting labels properly/moving labels from old index properly